src/org/sosy_lab/java_smt/solvers/cvc5/CVC5InterpolatingProver.java

```diff
@@ -0,0 +1,370 @@
+   * (1CVC5) and I -> B = True (2CVC5).
+   *
+   * <p>Craig Interpolation: There is a Model, such that the Interpolant psi, with phi- -> psi =
+   * True (1Craig), not(psi /\ phi+) = True (2Craig).
+   *
+   * <p>With A/phi- current set of assertions and B/phi+ Formulas to interpolate.
+   *
+   * <p>CVC5 -> Craig Interpolation:
+   *
+   * <p>(1CVC5) <=> (1Craig) due to subst of A with phi- and reflexivity.
+   *
+   * <p>(2CVC5) <=> I -> B = True <=> (not I) \/ B = True <=> not (I /\ (not B)) = True <=> (2Craig)
+   * due to subst of (not B) with phi+ and reflexivity.
+   *
+   * <p>Hence, CVC5's Interpolation produces an equivalent Interpolation to Craig Interpolation, if
+   * B is negated during CVC5 Interpolation.
+   *
+   * @param pFormAAsList current Set of Assertions A
+   * @param pFormBAsList Formulas to Interpolate B
+   * @return Interpolation of the Interpolation Pair following the definition of
+   *     Craig-Interpolation.
+   */
+  private Term getCVC5Interpolation(
+      ImmutableList<Collection<Term>> pFormAAsList, ImmutableList<Collection<Term>> pFormBAsList) {
+
+    // Respect Asserted Formulas not in the Interpolation Pairs
+
+    ImmutableList<Collection<Term>> combinedInterpols =
+        Stream.concat(pFormAAsList.stream(), pFormBAsList.stream())
+            .collect(ImmutableList.toImmutableList());
+
+    // checks, that no Assertions in the Assertion stack are left out. Can happen in Tree
+    // Interpolation, if pPartitionedFormulas does not contain every Set of Formulas from the
+    // Assertion stack.
+    Collection<Term> extraAssertions = getAssertedTermsNotInCollection(combinedInterpols);
+
+    // Uses a separate Solver instance to leave the original solver-context unmodified
+    Solver interpolationSolver = new Solver();
+
+    setSolverOptions(seed, solverOptions, interpolationSolver);
+
+    Term interpolantA = buildConjunctionOfFormulasOverList(pFormAAsList, interpolationSolver);
+
+    if (!extraAssertions.isEmpty()) {
+      Term extraAssert = buildConjunctionOfFormulas(extraAssertions, interpolationSolver);
+      interpolantA = interpolationSolver.mkTerm(Kind.AND, extraAssert, interpolantA);
+    }
+    Term interpolantB = buildConjunctionOfFormulasOverList(pFormBAsList, interpolationSolver);
+
+    interpolationSolver.assertFormula(interpolantA);
+
+    Term interpolant =
+        interpolationSolver.getInterpolant(interpolationSolver.mkTerm(Kind.NOT, interpolantB));
+
+    interpolationSolver.deletePointer();
+
+    return interpolant;
+  }
+
+  /**
+   * Returns Asserted Formulas (from the formula Stack), not present in the given collection of
+   * Terms.
+   *
+   * @param pCombinedInterpols asserted Formulas to invert the selection
+   * @return asserted Formulas not in collTerms, but in the formula Stack
+   */
+  private Collection<Term> getAssertedTermsNotInCollection(
+      ImmutableList<Collection<Term>> pCombinedInterpols) {
+    ImmutableList<Term> retTerms =
```

Comment:
Please consider runtime when using such data-structures.
A List requires a full iteration when searching for an element.
The code  few lines later checks for "contains". Very bad choice.

Tipp: You might want to use an ImmutableSet :-)

Reply:
Still have to learn about Java "types" ^^ I mostly replaced Immutable Lists with Immutable Sets.